test-case/check-volume-levels.sh

```diff
@@ -0,0 +1,386 @@
+    dlogi Set unmute
+    amixer cset name="$CAP_SWITCH" on
+}
+
+pause () {
+    dlogi Pausing
+    sleep 1
+}
+
+cset_mute_diff1 () {
+    dlogi Set switch pattern 1
+    case "$1" in
+	4)
+	    amixer cset name="$CAP_SWITCH" off,on,on,off
+	    ;;
+	2)
+	    amixer cset name="$CAP_SWITCH" off,on
+	    ;;
+	1)
+	    amixer cset name="$CAP_SWITCH" off
+	    ;;
+    esac
+}
+
+cset_mute_diff2 () {
+    dlogi Set switch pattern 2
+    case "$1" in
+	4)
+	    amixer cset name="$CAP_SWITCH" on,off,off,on
+	    ;;
+	2)
+	    amixer cset name="$CAP_SWITCH" on,off
+	    ;;
+	1)
+	    amixer cset name="$CAP_SWITCH" on
+	    ;;
+    esac
+}
+
+cset_volume_diff1 () {
+    dlogi Set volume pattern 1
+    case "$1" in
+	4)
+	    amixer cset name="$CAP_VOLUME" $VOL_P10DB,$VOL_0DB,$VOL_M10DB,$VOL_M30DB
+	    ;;
+	2)
+	    amixer cset name="$CAP_VOLUME" $VOL_P10DB,$VOL_0DB
+	    ;;
+	1)
+	    amixer cset name="$CAP_VOLUME" $VOL_P10DB
+	    ;;
+    esac
+}
+
+cset_volume_diff2 () {
+    dlogi Set volume pattern 2
+    case "$1" in
+	4)
+	    amixer cset name="$CAP_VOLUME" $VOL_M10DB,$VOL_P10DB,$VOL_0DB,$VOL_M20DB
+	    ;;
+	2)
+	    amixer cset name="$CAP_VOLUME" $VOL_M10DB,$VOL_P10DB
+	    ;;
+	1)
+	    amixer cset name="$CAP_VOLUME" $VOL_M10DB
+	    ;;
+    esac
+}
+
+arecord_helper () {
+    dlogi Capturing file "$1"
+    timeout -k 5 30 arecord -D "$CAP_HW" -f $CAP_FORMAT -r $CAP_RATE -c "$2" -d "$3" "$1"
+}
+
+aplay_helper () {
+    if ! [ -f "$1" ]; then
+	die "Error: File $1 does not exist"
+    fi
+
+    dlogi Playing file "$1"
+    timeout -k 5 120 aplay -D "$PLAY_HW" "$1" 
+}
+
+do_preparations () {
+
+    if [[ "$tplg" != *"nocodec"* ]]; then
+	dlogi "This test is executed only with nocodec topologies. Returning Not Applicable."
+	exit 2
+    fi
+
+    # Get max. channels count to use from capture device
+    # remove xargs after the trailing blank from tplgreader is fixed
+    CAP_CHANNELS=$($TPLGREADER "$tplg" -f "id:$PCM_ID & type:capture" -d ch_max -v | xargs)
+    CAP_HW=$($TPLGREADER "$tplg" -f "id:$PCM_ID & type:capture" -d dev -v | xargs)
+    PLAY_HW=$($TPLGREADER "$tplg" -f "id:$PCM_ID & type:playback" -d dev -v | xargs)
+    CAP_PGA=$($TPLGREADER "$tplg" -f "id:$PCM_ID & type:capture" -d pga -v | xargs)
+    PLAY_PGA=$($TPLGREADER "$tplg" -f "id:$PCM_ID & type:playback" -d pga -v | xargs)
+    export CAP_CHANNELS
+    export CAP_HW
+    export PLAY_HW
+
+    dlogi Test uses "$CAP_CHANNELS" channels
+    dlogi Playback device is "$PLAY_HW"
+    dlogi Playback PGA is "$PLAY_PGA"
+    dlogi Capture device is "$CAP_HW"
+    dlogi Capture PGA is "$CAP_PGA"
+
+    # Find amixer controls for capture, error if more than one PGA
+    numpga=${#CAP_PGA[@]}
+    if [ "$numpga" != 1 ]; then
+	die "Error: more than one capture PGA found."
+    fi
+
+    tmp=$(amixer controls | grep "$CAP_PGA" | grep Volume)
+    search="name="
+    CAP_VOLUME=${tmp#*$search}
+    export CAP_VOLUME
+
+    tmp=$(amixer controls | grep "$CAP_PGA" | grep Switch)
```

Comment:
This grep and above one for Volume need to be added || true to be able to next check for empty $tmp and in understandable way report what's the issue with topology.

Reply:
Done.